File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/modloaders/ForgeVersionListHandler.java
package net.kdt.pojavlaunch.modloaders;

import org.xml.sax.Attributes;
import org.xml.sax.SAXException;
import org.xml.sax.helpers.DefaultHandler;

import java.util.ArrayList;
import java.util.List;

/**
 * ForgeのメタデータXMLをパースするSAXハンドラー
 */
public class ForgeVersionListHandler extends DefaultHandler {
    private List<String> mForgeVersions;
    private StringBuilder mCurrentVersion = null;

    /**
     * ドキュメント解析開始時にバージョンリストを初期化します。
     */
    @Override
    public void startDocument() throws SAXException {
        mForgeVersions = new ArrayList<>();
    }

    /**
     * XML要素内のテキストを収集します。
     */
    @Override
    public void characters(char[] ch, int start, int length) throws SAXException {
        if(mCurrentVersion != null) mCurrentVersion.append(ch, start, length);
    }

    /**
     & version要素の開始時に新しいStringBuilderを作成します。
     */
    @Override
    public void startElement(String uri, String localName, String qName, Attributes attributes) throws SAXException {
        if(qName.equals("version")) mCurrentVersion = new StringBuilder();
    }

    /**
     * version要素の終了時に収集したバージョンをリストに追加します。
     */
    @Override
    public void endElement(String uri, String localName, String qName) throws SAXException {
        if (qName.equals("version")) {
            String version = mCurrentVersion.toString();
            mForgeVersions.add(version);
            mCurrentVersion = null;
        }
    }

    /**
     * 解析結果のバージョンリストを返します。
     */
    public List<String> getVersions() {
        return mForgeVersions;
    }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/feature/mod/modloader/ForgeVersionListHandler.kt
package com.arata.yukarilauncher.feature.mod.modloader

import org.xml.sax.Attributes
import org.xml.sax.helpers.DefaultHandler

/** ForgeのMavenメタデータXMLをパースしてバージョン一覧を抽出するSAXハンドラ。 */
class ForgeVersionListHandler : DefaultHandler() {
    /** パース中のForgeバージョンリスト。 */
    private var mForgeVersions: MutableList<String>? = null
    /** 現在読み込み中のバージョン文字列バッファ。 */
    private var mCurrentVersion: StringBuilder? = null

    /** XMLドキュメントの開始時にバージョンリストを初期化する。 */
    override fun startDocument() {
        mForgeVersions = ArrayList()
    }

    /** XML要素内の文字データをバッファに追加する。 @param ch 文字配列 @param start 開始位置 @param length 長さ */
    override fun characters(ch: CharArray, start: Int, length: Int) {
        if (mCurrentVersion != null) mCurrentVersion!!.append(ch, start, length)
    }

    /** XML要素の開始タグを処理する。"version"タグで新しいバッファを作成する。 @param uri 名前空間URI @param localName ローカル名 @param qName 修飾名 @param attributes 属性 */
    override fun startElement(uri: String?, localName: String?, qName: String?, attributes: Attributes?) {
        if (qName == "version") mCurrentVersion = StringBuilder()
    }

    /** XML要素の終了タグを処理する。"version"タグでバージョンをリストに追加する。 @param uri 名前空間URI @param localName ローカル名 @param qName 修飾名 */
    override fun endElement(uri: String?, localName: String?, qName: String?) {
        if (qName == "version") {
            val version = mCurrentVersion.toString()
            mForgeVersions!!.add(version)
            mCurrentVersion = null
        }
    }

    /** パース済みのForgeバージョンリストを取得する。 @return バージョン文字列のリスト */
    val versions: List<String>
        get() = mForgeVersions!!
}